### ISSUE TITLE
[🐛 Bug]: Explicit Wait Affected by Implicit Wait 

### ISSUE THREAD

**Issue by braimanm** [SeleniumHQ/selenium#14635]
### Description:
In Selenium, both explicit wait (WebDriverWait or FluentWait) and implicit wait are designed to control how long WebDriver waits for elements to appear in the DOM before throwing an exception. However, there is currently a dependency between these two mechanisms, which leads to inefficient behavior.

Specifically, when using explicit wait with ExpectedConditions (like visibilityOfElementLocated() and others), **the findElement() method is called internally**. Since **findElement()** is affected by implicit wait, this results in a conflict when the implicit wait duration is longer than the explicit wait timeout. This causes WebDriverWait to wait for an undefined amount of time, influenced by the implicit wait, and leads to incorrect exception messages being generated.

### Proposed Solution:
To improve the precision and efficiency of waits in Selenium, the ExpectedConditions class methods should be decoupled from the global implicit wait by not invoking **_findElement()_** or **_findElements()_** methods but instead using other mechanism which is not affected by implicit wait. This would ensure that explicit waits such as WebDriverWait or FluentWait work independently of the implicit wait setting, giving test developers finer control over waiting behavior.

### How can we reproduce the issue?

```java
WebDriver driver = new ChromeDriver();
driver.manage().timeouts().implicitlyWait(30, TimeUnit.SECONDS);
driver.get("https://uitaf.org/test/");
WebDriverWait wait = new WebDriverWait(driver, Duration.ofSeconds(5));
WebElement element = wait.until(ExpectedConditions.visibilityOfElementLocated(By.id("non-existing")));

```

### Relevant log output

`
org.openqa.selenium.TimeoutException: Expected condition failed: waiting for visibility of element located by By.id: non-existing (tried for 5 second(s) with 500 milliseconds interval)
`

> “The log incorrectly shows that we waited for 5 seconds, but in reality, the implicit wait affected the total time, and we actually waited for 30+ seconds.”

### Operating System

na

### Selenium version

na

### What are the browser(s) and version(s) where you see this issue?

na

### What are the browser driver(s) and version(s) where you see this issue?

na

### Are you using Selenium Grid?

na

**Comment by braimanm** [comment]
I consulted ChatGPT regarding the question: **‘How can I find an element without being affected by implicit waits?’** 
The response provided aligns with standard Selenium practices:
_**"To bypass the influence of implicit waits when locating an element, you can utilize explicit waits. Below is a guide on implementing this in Selenium: …"**_

However, as I noted, this approach is not considered best practice in this context.

**Comment by shbenzer** [comment]
This conflict has been present for some time - it is mentioned in the selenium docs as a warning https://www.selenium.dev/documentation/webdriver/waits/#implicit-waits

**Comment by braimanm** [comment]
Thank you for the reference. This issue has persisted for years, isn’t now a good time to fix it? It creates a status quo where explicit waits are unreliable. Since Selenium is an API used by various frameworks to simplify test automation for non-professional developers, this conflict introduces significant challenges. Alternatively, it may be worth considering the removal of implicit waits altogether.

**Comment by diemol** [comment]
I am not following. Why is this an issue? The Selenium project has advised over several years to not mix implicit and explicit waits. 

When mixed, we cannot simply deactivate one or the other, mostly because implicit waiting is done on the server side (ChromeDriver, GeckoDriver, etc.), not in the Selenium code itself. Hence the advice.

I understand the frustration, but we won't make any changes related to that given our clarity on the advice over years.

I will close this issue but feel free to keep the conversation going.

**Comment by braimanm** [comment]
@diemol 
As an active member of numerous Selenium-based and other test automation LinkedIn groups, I frequently observe a consistent issue regarding the understanding of Implicit Waits and Explicit Waits. In many code examples that I’ve reviewed especially those written in Java there is often a mix of both Implicit and Explicit waits, which leads to confusion.

To address this issue in Java, the solution is straightforward. You can fix it by modifying the **FluentWait** class’s **until(Function<? super T, V> isTrue)** method. The following logic ensures that Implicit Wait and Explicit Wait do not conflict:

```java
// Save the current implicit wait duration 
Duration duration = driver.manage().timeouts().getImplicitWaitTimeout();
// Set implicit wait to 0
driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(0));
// Execute the until method logic
// Before returning from the until method, reset the implicit wait to the originally saved value
// This should also be done in all catch blocks before throwing TimeoutException or WebDriverException
driver.manage().timeouts().implicitlyWait(duration);
```

**Comment by MJB222398** [comment]
@braimanm You're not supposed to mix implicit and explicit waits. When you see them being mixed together why do you want to keep both and try to make it work? Why are you not fixing it by removing one or the other? Selenium is very clear that they should not be used together. Being able to read documentation that says "Do NOT use both these things together" is not what separates professional developers from non-professional developers/laymen. If you are working with people that cannot read and understand that and you cannot educate them so they stop doing this then it is most definitely those people that are your problem. Surely you cannot expect Selenium to "fix" this when 99.99% of people can understand this completely logical limitation?

**Comment by braimanm** [comment]
Hi @MJB222398, sorry that I can’t call you by your name. As I mentioned before, I’m part of many Test Automation communities and an active contributor. Being a professional developer is not only about understanding how specific software works but also about recognizing patterns in misunderstandings and helping others. For example, I can show you several scenarios where Implicit and Explicit waits can work together, but the Selenium team decided not to fix this issue, opting instead to add a warning in the documentation. This is what makes us human and fuels our passion to help others.

**Rationale for the fix:** Many frameworks are built with default behaviors to handle common scenarios, while still allowing developers the flexibility to temporarily override these settings when necessary. This design approach also applies effectively to managing waits.
